accounts-db/src/accounts_db.rs

```diff
@@ -6295,33 +6295,42 @@ impl AccountsDb {
         });
 
         // Always flush up to `requested_flush_root`, which is necessary for things like snapshotting.
-        let cached_roots: BTreeSet<Slot> = self.accounts_cache.clear_roots(requested_flush_root);
+        let flushed_roots: BTreeSet<Slot> = self.accounts_cache.clear_roots(requested_flush_root);
 
         // Iterate from highest to lowest so that we don't need to flush earlier
         // outdated updates in earlier roots
         let mut num_roots_flushed = 0;
         let mut flush_stats = FlushStats::default();
-        for &root in cached_roots.iter().rev() {
+        for &root in flushed_roots.iter().rev() {
             if let Some(stats) =
                 self.flush_slot_cache_with_clean(root, should_flush_f.as_mut(), max_clean_root)
             {
                 num_roots_flushed += 1;
                 flush_stats.accumulate(&stats);
             }
 
-            // Regardless of whether this slot was *just* flushed from the cache by the above
-            // `flush_slot_cache()`, we should update the `max_flush_root`.
-            // This is because some rooted slots may be flushed to storage *before* they are marked as root.
-            // This can occur for instance when
-            //  the cache is overwhelmed, we flushed some yet to be rooted frozen slots
-            // These slots may then *later* be marked as root, so we still need to handle updating the
-            // `max_flush_root` in the accounts cache.
-            self.accounts_cache.set_max_flush_root(root);
+            if let Some(storage) = self.storage.get_slot_storage_entry(root) {
+                // In this loop, "flush_slot_cache_with_clean" will handle
+                // cleaning the accounts for only the "flushed_roots". However,
+                // these "flushed_roots" may trigger additional clean for older
+                // roots. Therefore, we need to add them to "dirty_stores" for
+                // clean to pick up such additional cleaning.
+                self.dirty_stores.entry(root).or_insert(storage);
+            }
         }
 
-        // Only add to the uncleaned roots set *after* we've flushed the previous roots,
-        // so that clean will actually be able to clean the slots.
-        let num_new_roots = cached_roots.len();
+        // Regardless of whether this slot was *just* flushed from the cache by
```

Comment:
An optimization to move the "max_flush_root" update out of the loop. 
We only need to fetch_max with the max value in the "flushed_roots".

Reply:
This code is introduced 4 years ago, and has never changed since them. @carllin 

https://github.com/solana-labs/solana/pull/14596/commits/8f7c7fbd1011d39d017e7073692ba74894175888#diff-1090394420d51617f3233275c2b65ed706b35b53b115fe65f82c682af8134a6fR3017

@carllin Is there any reason that we can't move `set_max_flush_root` out of the loop?